<!DOCTYPE html>
<html>
<head>
    <title>Converted from: projectscope.md</title>
    <meta charset="utf-8"/>

</head>
<body>

<div
            class="github-markdown-body"
            data-color-mode="auto"
            data-light-theme="light"
            data-dark-theme="dark"
        >
            <div class="github-markdown-content"><span id="markdown-mermaid" aria-hidden="true"
                    data-dark-mode-theme="dark"
                    data-light-mode-theme="default"></span>
                <h2 id="heading-background3a">Background:</h2>
<p>In a previous study, heterogeneity exploration of bulk transcriptomic profiling was visually and quantitatively compared by applying PCA, MDS, t-SNE, and UMAP when parameters were set as default. In 71 sizable datasets of bulk transcriptomic profiling in two-dimensional space, UMAP was shown to be superior in preserving sample-level neighborhood information and maintaining clustering accuracy, thus conspicuously differentiating batch effects, identifying pre-defined biological groups, and identifying clustering structures associated with biological features and clinical meaning.</p><p>However, clustering suffers from both the limitations of the algorithms and the Curse of Dimensionality, which often results in a less consensus between the visual interpretation of the tSNE plot and the clusters generated. In the absence of hyper-parameter optimization performed on the clustering models or for the embedding spaces generated by dimensionality reduction techniques (t-SNE, UMap, etc), the results can be unreliable due to reproducibility issues.</p><p>Paper: <a href="https://www.sciencedirect.com/science/article/pii/S2211124721008597">Dimensionality reduction Application by UMAP in bulk transcriptomic data</a></p><h2 id="heading-problem-statement">Problem statement</h2>
<p>Given the unpredictable nature of clustering algorithms and often non-intuitive nature of a-priori expectation of an embedding manifold, we want to investigate the effect on hyper-parameter optimization on cluster models and embedding space of UMap for better stratification of cohorts in bulk-transcriptomics datasets.</p><h4 id="heading-goals3a">Goals:</h4>
<p>For ~1500 samples total for 3 sub-types of Kidney cancer and 60660 gene expression features (Normalized fragments per kilo million[fpkm]), how well can we stratify the sub-types and quantify genes associated with each cohort?</p><h6 id="heading-a.-sub-type-stratification3a">A. Sub-type stratification:</h6>
<p>A.1. Visual results:</p><ul data-line="19" class="code-line" dir="auto">
<li>How well do default embeddings of UMAP in euclidean and non-euclidean spaces separate the 3 cohorts?</li>
<li>How does hyper-parameter optimized model of UMAP + clustering compare in cohort stratification compared to default embeddings?</li>
<li>Which method provides best overall stratification of the sub-types visually?</li>
<li>Which method shows better segregation of batch effects?</li>
</ul>
<p>A.2. Performance results:</p><ul data-line="25" class="code-line" dir="auto">
<li><p>What is the performance of different clustering methods on DR projections?</p><ul data-line="26" class="code-line" dir="auto">
<li><p>Metrics for measuring performance of clustering methods based on the paper:</p><ol data-line="28" class="code-line" dir="auto">
<li>Clustering accuracy</li>
<li>Neighborhood preserving</li>
<li>Computational efficiency</li>
</ol>
</li>
</ul>
</li>
<li><p>How does best UMAP model compare to other DR methods (t-SNE, Local linear embedding, etc) in terms of supervised learning performance?</p><ul data-line="33" class="code-line" dir="auto">
<li><p>Metrics:</p><ul data-line="35" class="code-line" dir="auto">
<li>Accuracy</li>
<li>Precision</li>
<li>Recall</li>
</ul>
</li>
</ul>
</li>
<li><p>How good is the biological interpretability of clusters formed compared to other methods (t-SNE, PCA, spectral clustering, etc)?</p></li>
</ul>
<h6 id="heading-b.-quantification-of-genes-associated-with-sub-types3a">B. Quantification of genes associated with sub-types:</h6>
<ul data-line="44" class="code-line" dir="auto">
<li>What genes are up-regulated or down-regulated in one sub-chort vs another? i.e what are the differential gene expressions between 2 cohorts?</li>
</ul>
<h2 id="heading-recommended-umap-pipeline">Recommended UMAP Pipeline</h2>
<p><img src="images/PipelineForUmapBulkRNA.jpg" alt="" data-src="images/PipelineForUmapBulkRNA.jpg" />
<a href="https://www.sciencedirect.com/science/article/pii/S2211124721008597#sec4">Paper reference</a> Yang Yang, Hongjian Sun, Yu Zhang, Tiefu Zhang, Jialei Gong, Yunbo Wei, Yong-Gang Duan, Minglei Shu, Yuchen Yang, Di Wu, Di Yu,
Dimensionality reduction by UMAP reinforces sample heterogeneity analysis in bulk transcriptomic data,
Cell Reports,
Volume 36, Issue 4,
2021,
109442,
ISSN 2211-1247,
<a href="https://doi.org/10.1016/j.celrep.2021.109442">https://doi.org/10.1016/j.celrep.2021.109442</a>.</p><h2 id="heading-methods2fsoftware-utilities">Methods/Software Utilities</h2>
<h4 id="heading-software-utilities">Software utilities</h4>
<h6 id="heading-i.-pre-processing-utililies-for-dr3a">I. Pre-Processing utililies for DR:</h6>
<ul data-line="63" class="code-line" dir="auto">
<li>Transformations:
<ul data-line="64" class="code-line" dir="auto">
<li>Using standard scaler on fpkm_unstranded values for tightening the range of values.</li>
<li>Using log transformation</li>
<li>Variance stabilized transformation</li>
</ul>
</li>
<li>Using 3 different types of mappers (plane, sphere and hyperbeloid), visualize the reduced dataset for Class Labels TCGA-KICH, TCGA-KIRC, TCGA-KIRP</li>
</ul>
<h6 id="heading-ii.-list-of-dr-to-be-explored">II. List of DR to be explored</h6>
<ul data-line="71" class="code-line" dir="auto">
<li>Parametric (neural network) Embedding</li>
<li>t-SNE</li>
<li>PCA</li>
<li>MDS</li>
<li>UMap</li>
</ul>
<h6 id="heading-iii.-list-of-clustering-methods-to-be-applied3a">III. List of clustering methods to be applied:</h6>
<ul data-line="79" class="code-line" dir="auto">
<li>K-means</li>
<li>Hierarchical</li>
<li>Spectral</li>
<li>HDBScan</li>
<li>GMM</li>
<li>Possibly deep learning methods? <a href="https://academic.oup.com/bib/article/22/1/393/5721075">Review Paper link</a></li>
</ul>
<p>A list of 10 of the more popular algorithms is as follows:
(<a href="https://machinelearningmastery.com/clustering-algorithms-with-python/">Examples of clustering methods applied</a>)</p><ul data-line="89" class="code-line" dir="auto">
<li>Affinity Propagation</li>
<li>Agglomerative Clustering</li>
<li>BIRCH</li>
<li>DBSCAN</li>
<li>K-Means</li>
<li>Mini-Batch K-Means</li>
<li>Mean Shift</li>
<li>OPTICS</li>
<li>Spectral Clustering</li>
<li>Mixture of Gaussians</li>
</ul>
<h4 id="heading-method-details3a">Method details:</h4>
<h6 id="heading-i.-investigation-of-group-separation-by-different-umap-models.">I. Investigation of group separation by different UMAP models.</h6>
<p>Objective: Given the raw RNA-Seq gene expression data of kidney cancer in 3 sub-types, can a simple projection of the high-dimensional features into euclidean or non-euclidean space give good cohort stratification?</p><p>Models used to generate embeddings:</p><ol data-line="105" class="code-line" dir="auto">
<li><p>Umap with default config - This generates embeddings in the euclidean space</p></li>
<li><p>Umap with  <code class="hljs"><span class="hljs-attribute">Hyperbeloid</span></code> output metric - Generates a non-euclidean space embedding.</p></li>
</ol>
<h6 id="heading-ii.-investigation-of-more-optimized-and-well-separated-clustering-structures-in-kidney-cancer">II. Investigation of more optimized and well separated clustering structures in kidney cancer</h6>
<h6 id="heading-objective3a">Objective:</h6>
<p>Given that initial Umap embeddings did not capture separation of clusters to provide any strong insight into underlying biological or batch effects, we want to apply a bayesian hyper parameter optimization strategy for UMAP embeddings + DBSCAN clustering to get  optimal parameters for UMAP and DBSCAN.</p><h6 id="heading-motivation-for-using-hdbscan-over-other-clustering-models3a">Motivation for using HDbScan over other clustering models:</h6>
<p>HDBSCAN, i.e. Hierarchical DBSCAN, is a powerful density-based clustering algorithm which is:</p><ol data-line="116" class="code-line" dir="auto">
<li>indifferent to the shape of clusters,</li>
<li>does not require the number of clusters to be specified,</li>
<li>robust with respect to clusters with different density.</li>
<li>Further, HBDSCAN is very attractive because it has only one hyperparameter minimum cluster size which is the minimal number of points in a cluster.</li>
<li>It is relatively fast for large data sets, detects outlying cells, and for each cell it reports a probability of assignment to a cluster</li>
</ol>
<h6 id="heading-evaluation-of-best-clusters-to-select-the-best-hyperparameters3a">Evaluation of best clusters to select the best hyperparameters:</h6>
<ul data-line="124" class="code-line" dir="auto">
<li><p>Silhouette Score is not a good validation metric for density-based algorithms like DBSCAN and HDBSCAN since it assumes all points are assigned a group and can’t appropriately handle noise/outliers that are seen in the initial clustering of our UMAP embeddings at the intersection of different project (Fig 1A, 1B and 1C).</p></li>
<li><p>Density Based Cluster Validation (DBCV) has been proposed and used by some for tuning HDBSCAN hyperparameters. While it likely works well for several applications, it can favor having a smaller number of clusters and leaving too many samples in the “noise” category</p></li>
<li><p>Here we will leverage <code class="hljs">HDbScan<span class="hljs-symbol">&#x27;s</span></code> probabilities attribute.</p><ul data-line="129" class="code-line" dir="auto">
<li>Definition from documentation: The strength with which each sample is a member of its assigned cluster. Noise points have probability zero; points in clusters have values assigned proportional to the degree that they persist as part of the cluster</li>
</ul>
</li>
</ul>
<p>Cost function = percent of dataset with &lt; 10% cluster label confidence</p><h2 id="heading-visualizations2fresults3a">Visualizations/Results:</h2>
<p>This section has the visualization of sub-cohorts after DR, interpretation of results and visualization of quantifying metrics for measuring clustering accuracy, neighborhood preservation and classification accuracy.</p><h4 id="heading-i.-investigation-of-group-separation-by-different-umap-embeddings3a">I. Investigation of group separation by different UMAP embeddings:</h4>
<p>Performance of UMAP as a function of different embedding spaces:</p><h6 id="heading-umap-using-default-configurations3a">Umap using default configurations:</h6>
<p>Fig 1.A. Plane embedding shows separation of clustering structures by project_id<img src="/UmapApplication/images/Plane_UMAP_embedment.png" alt="2D scatterplot using plane mapper" data-src="/UmapApplication/images/Plane_UMAP_embedment.png" /></p><p>Fig 1.B. Hyperbeloid embedding supports the clustering by project_id <img src="images/Hperbolic_UMAP_embedder.png" alt="2D scatterplot using hyperbeloid mapper" data-src="images/Hperbolic_UMAP_embedder.png" /></p><div  data-line="149" class="code-line" dir="auto" ></div>
<!-- Fig 1.C. Example of 3D visualization shows separate projects by clustering structures   ![3D scatterplot using Sphere mapper](images/Sphere_UMAP_embedment.png) -->
<h6 id="heading-interpretation3a">Interpretation:</h6>
<p>In the 2-D plots, there are 4 distinct clusters that are observed. 3 of the 4 clusters have majority of the class labels being well separated. However, there is one cluster that seems mixed with all three sub-types of kidney cancer.</p><ul data-line="156" class="code-line" dir="auto">
<li>The well separater Orange, Green and Blue clusters of TCGA-KIRP, TCGA-KICH and TCGA-KIRC cohorts seem to be from 3 different projects.</li>
<li>However the overall separation boundary between projects is not as distinct.</li>
<li>Additionally, The mixed 4th group might correspond to a different biological group.</li>
</ul>
<h5 id="heading-ii.-investigation-of-more-optimized-and-well-separated-clustering-structures-in-kidney-cancer.">II. Investigation of more optimized and well separated clustering structures in kidney cancer</h5>
<p>Fig 2.A. Separation of clustering structures with bayesian optimized HDbScan<img src="/UmapApplication/images/OptimizedClusteringGE.png" alt="Separation of clustering structures with bayesian optimized HDbScan" data-src="/UmapApplication/images/OptimizedClusteringGE.png" /></p><h6 id="heading-interpretation3a.">Interpretation:</h6>
<ol data-line="168" class="code-line" dir="auto">
<li><p>In the 3 plots in figure 2.A, bayesian optimized Umap Cosine embedding with HDbScan clustering leads to identification of 6 well separated clusters.</p><ol data-line="169" class="code-line" dir="auto">
<li>4 large clusters (0, 2,3 and 5) are seen.</li>
<li>2 small clusters (1 and 4) are seen.</li>
</ol>
</li>
<li><p>The 4 large clusters are similar to the 4 large clusters seen in figure 1.A and 1.B, but with better separation.</p><ol data-line="173" class="code-line" dir="auto">
<li>The large clusters (0, 2, 3) are representative of the 3 kidney cancer subtypes (KIRC, KIRP and KICH)</li>
<li>The large cluster (5) comprised of mixed populations of the 3 kidney cnacer subtypes are representative of a &quot;SOLID TISSUE NORMAL&quot; Tumor as compared to &quot;PRIMARY TUMOR&quot; in clusters (0,2,3)</li>
</ol>
</li>
<li><p>Overall, the bayesian optimized Umap HdbScan is able to improve separability of clustering structures based on biological background.</p></li>
</ol>
<h4 id="heading-next-steps3a">NEXT STEPS:</h4>
<p>Evaluation of Classification Performance results:</p><ul data-line="182" class="code-line" dir="auto">
<li><p>What is the performance of different clustering methods on DR projections?</p><ul data-line="183" class="code-line" dir="auto">
<li><p>Metrics for measuring performance of clustering methods based on the paper:</p><ol data-line="185" class="code-line" dir="auto">
<li>Clustering accuracy</li>
<li>Neighborhood preserving</li>
<li>Computational efficiency</li>
</ol>
</li>
</ul>
</li>
<li><p>How does best UMAP model compare to other DR methods (t-SNE, Local linear embedding, etc) in terms of supervised learning performance?</p><ul data-line="190" class="code-line" dir="auto">
<li><p>Metrics:</p><ul data-line="192" class="code-line" dir="auto">
<li>Accuracy</li>
<li>Precision</li>
<li>Recall</li>
</ul>
</li>
</ul>
</li>
<li><p>How good is the biological interpretability of clusters formed compared to other methods (t-SNE, PCA, spectral clustering, etc)?</p></li>
</ul>
</div>
        </div>

</body>
</html>
