<!DOCTYPE html>
<html>
<head>
    <title>Converted from: workflow.md</title>
    <meta charset="utf-8"/>

</head>
<body>

<div
            class="github-markdown-body"
            data-color-mode="auto"
            data-light-theme="light"
            data-dark-theme="dark"
        >
            <div class="github-markdown-content"><span id="markdown-mermaid" aria-hidden="true"
                    data-dark-mode-theme="dark"
                    data-light-mode-theme="default"></span>
                <script type="module" src="https://unpkg.com/@github/clipboard-copy-element@latest"></script>
<style>
    .ClipboardButton {
        position: relative;
    }

    clipboard-copy {
        /* padding: 0.4em 0.6em; */
        font: 0.9rem system-ui, sans-serif;
        display: inline-block;
        cursor: default;
        color: var(--vscode-button-foreground);
        background-color: var(--vscode-button-border);
        border-color: var(--vscode-textCodeBlock-background);
        border-radius: 6px;
        border: 1px solid var(--vscode-widget-border);
        box-shadow: inset 0 -1px 0 var(--vscode-widget-shadow);
        transition: 80ms cubic-bezier(0.33, 1, 0.68, 1);
        transition-property: color,background-color,box-shadow,border-color;
    }

    clipboard-copy:hover {
        background-color: var(--vscode-statusBarItem-hoverBackground);
    }

    clipboard-copy:active {
        outline: 1px solid var(--vscode-contrastActiveBorder) !important;
        outline-offset: -1px;
    }

    clipboard-copy:focus-visible {
        outline: 1px solid var(--vscode-focusBorder);
    }

    .d-none {
        display: none !important;
    }

    .position-relative {
        position: relative !important;
    }

    .position-absolute {
        position: absolute !important;
    }

    .right-0 {
        right: 0 !important;
    }

    .top-0 {
        top: 0 !important;
    }

    .m-2 {
        margin: var(--base-size-8, 8px) !important;
    }

    .color-fg-success {
        color: #3fb950 !important;
        stroke: #3fb950 !important;
        border-color: #3fb950 !important;
    }
</style>
<script>
    document.addEventListener('clipboard-copy', function (event) {
        const notice_clipboard_copy = event.target;
        console.log(`notice_clipboard_copy = ${notice_clipboard_copy}`)
        const notice_copy = event.target.querySelector('.clipboard-copy-icon');
        const notice_check = event.target.querySelector('.clipboard-check-icon');
        notice_clipboard_copy.classList.add("color-fg-success");
        notice_copy.classList.add("d-none");
        notice_check.classList.remove("d-none");
        setTimeout(function () {
            notice_clipboard_copy.classList.remove("color-fg-success");
            notice_copy.classList.remove("d-none");
            notice_check.classList.add("d-none");
        }, 1000)
    })
</script><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>import OmicsUtils.DimRedMappers.clusterer
import OmicsUtils.DimRedMappers.umap_embedders
from importlib import reload
reload(OmicsUtils.DimRedMappers.clusterer)
reload(OmicsUtils.DimRedMappers.umap_embedders)

import pandas as pd 
import numpy as np
from sklearn.preprocessing import FunctionTransformer
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import seaborn as sns
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="import OmicsUtils.DimRedMappers.clusterer
import OmicsUtils.DimRedMappers.umap_embedders
from importlib import reload
reload(OmicsUtils.DimRedMappers.clusterer)
reload(OmicsUtils.DimRedMappers.umap_embedders)

import pandas as pd 
import numpy as np
from sklearn.preprocessing import FunctionTransformer
from sklearn.model_selection import train_test_split
import matplotlib.pyplot as plt
import seaborn as sns" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="15" class="code-line" dir="auto"><code class="hljs">/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><h2 id="heading-i.-data-preprocessing-and-transformation">I. Data Preprocessing and transformation</h2>
<h2 id="heading-1.1-read-and-wrangle-the-data-to-get-gene-expression-features-and-clinical-data-into-a-single-dataframe">1.1 Read and wrangle the data to get gene expression features and clinical data into a single dataframe</h2>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>## Read labels and gene expression data 
fpkm_unstr_df_with_labels = pd.read_csv('./Transcriptomics/data/processed_data/fpkm_unstr_data_with_labels.csv')
gene_cols = fpkm_unstr_df_with_labels.columns.to_numpy()[:60660]
exposure_tsv = pd.read_csv('./Transcriptomics/data/clinical.cart.2023-10-29/exposure.tsv', sep='\t')


## some columns have whitespace that needs to be removed for better processing
new_columns = fpkm_unstr_df_with_labels.columns.str.replace(' ', '_').to_numpy()
fpkm_unstr_df_with_labels.columns = new_columns 

## Merge exposure data to get labels for tumor type and kidney subtype
ge_kidney_cancer_data_with_tgca_labels = pd.merge(fpkm_unstr_df_with_labels,
                                                  exposure_tsv[['case_submitter_id', 'project_id']],
                                                  left_on='Case_ID',
                                                  right_on='case_submitter_id')
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="## Read labels and gene expression data 
fpkm_unstr_df_with_labels = pd.read_csv(&#039;./Transcriptomics/data/processed_data/fpkm_unstr_data_with_labels.csv&#039;)
gene_cols = fpkm_unstr_df_with_labels.columns.to_numpy()[:60660]
exposure_tsv = pd.read_csv(&#039;./Transcriptomics/data/clinical.cart.2023-10-29/exposure.tsv&#039;, sep=&#039;\t&#039;)


## some columns have whitespace that needs to be removed for better processing
new_columns = fpkm_unstr_df_with_labels.columns.str.replace(&#039; &#039;, &#039;_&#039;).to_numpy()
fpkm_unstr_df_with_labels.columns = new_columns 

## Merge exposure data to get labels for tumor type and kidney subtype
ge_kidney_cancer_data_with_tgca_labels = pd.merge(fpkm_unstr_df_with_labels,
                                                  exposure_tsv[[&#039;case_submitter_id&#039;, &#039;project_id&#039;]],
                                                  left_on=&#039;Case_ID&#039;,
                                                  right_on=&#039;case_submitter_id&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><h2 id="heading-1.2.-data-transformation">1.2. Data transformation</h2>
<p>Here, given we have RNA-Seq Gene expression data we apply some commonly applied data transformations</p><ol data-line="46" class="code-line" dir="auto">
<li><strong>Log transformation</strong>: Gene expression values have a very wide range from (0, 500000) usually in the form of a spike-slab distribution. Hence need to tighten the range with a log transform</li>
<li><strong>Subsetting</strong>: For purposes of this analysis we only want to look at 3 sub-types of kidney cancer samples ('TCGA-KIRC', 'TCGA-KICH', 'TCGA-KIRP')</li>
<li><strong>Label conversion</strong>: Converting categorical labels into numerical labels for compatibility with machine learning algorithms</li>
</ol>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>description_df = fpkm_unstr_df_with_labels.iloc[:, :60660].describe().reset_index()
showing_value_range = description_df[description_df['index'].isin(['min', 'mean', 'max'])].iloc[:,1:].values.flatten()

</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="description_df = fpkm_unstr_df_with_labels.iloc[:, :60660].describe().reset_index()
showing_value_range = description_df[description_df[&#039;index&#039;].isin([&#039;min&#039;, &#039;mean&#039;, &#039;max&#039;])].iloc[:,1:].values.flatten()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax = sns.distplot(showing_value_range)
ax.set_title("Distribution of min, mean, max of all gene expression values" )
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax = sns.distplot(showing_value_range)
ax.set_title(&quot;Distribution of min, mean, max of all gene expression values&quot; )" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="63" class="code-line" dir="auto"><code class="hljs">/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/796282352.py:1: UserWarning: 

`distplot` is a deprecated function and will be removed in seaborn v0.14.0.

Please adapt your code to use either `displot` (a figure-level function with
similar flexibility) or `histplot` (an axes-level function for histograms).

For a guide to updating your code to use the new functions, please see
https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751

  ax = sns.distplot(showing_value_range)





Text(0.5, 1.0, 'Distribution of min, mean, max of all gene expression values')
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/796282352.py:1: UserWarning: 

`distplot` is a deprecated function and will be removed in seaborn v0.14.0.

Please adapt your code to use either `displot` (a figure-level function with
similar flexibility) or `histplot` (an axes-level function for histograms).

For a guide to updating your code to use the new functions, please see
https://gist.github.com/mwaskom/de44147ed2974457ad6372750bbe5751

  ax = sns.distplot(showing_value_range)





Text(0.5, 1.0, &#039;Distribution of min, mean, max of all gene expression values&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_6_2.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_6_2.png" /></p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>
## Subset only 3 types of cancer sub-type from the dataset 
ge_kidney_cancer_data_correct_labels = (ge_kidney_cancer_data_with_tgca_labels[ge_kidney_cancer_data_with_tgca_labels['project_id'].isin(
    ['TCGA-KIRC', 'TCGA-KICH', 'TCGA-KIRP']
)]
)

## Apply log transformation
transformer = FunctionTransformer(np.log10)
ge_kidney_cancer_data_correct_labels[gene_cols] = ge_kidney_cancer_data_correct_labels[gene_cols] + 1
ge_kidney_cancer_data_correct_labels[gene_cols] = transformer.fit_transform(ge_kidney_cancer_data_correct_labels[gene_cols])

## Convert categorical labels to numerical labels 
columns_for_one_hot = ['project_id', 'Sample_Type']
ml_df = pd.get_dummies(ge_kidney_cancer_data_correct_labels, columns=['project_id', 'Sample_Type'], prefix=['cancer_subtype', 'tumor_subtype'])
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="## Subset only 3 types of cancer sub-type from the dataset 
ge_kidney_cancer_data_correct_labels = (ge_kidney_cancer_data_with_tgca_labels[ge_kidney_cancer_data_with_tgca_labels[&#039;project_id&#039;].isin(
    [&#039;TCGA-KIRC&#039;, &#039;TCGA-KICH&#039;, &#039;TCGA-KIRP&#039;]
)]
)

## Apply log transformation
transformer = FunctionTransformer(np.log10)
ge_kidney_cancer_data_correct_labels[gene_cols] = ge_kidney_cancer_data_correct_labels[gene_cols] + 1
ge_kidney_cancer_data_correct_labels[gene_cols] = transformer.fit_transform(ge_kidney_cancer_data_correct_labels[gene_cols])

## Convert categorical labels to numerical labels 
columns_for_one_hot = [&#039;project_id&#039;, &#039;Sample_Type&#039;]
ml_df = pd.get_dummies(ge_kidney_cancer_data_correct_labels, columns=[&#039;project_id&#039;, &#039;Sample_Type&#039;], prefix=[&#039;cancer_subtype&#039;, &#039;tumor_subtype&#039;])" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="108" class="code-line" dir="auto"><code class="hljs">/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/1726230766.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels[gene_cols] = ge_kidney_cancer_data_correct_labels[gene_cols] + 1
/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/1726230766.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels[gene_cols] = transformer.fit_transform(ge_kidney_cancer_data_correct_labels[gene_cols])
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/1726230766.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels[gene_cols] = ge_kidney_cancer_data_correct_labels[gene_cols] + 1
/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/1726230766.py:10: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels[gene_cols] = transformer.fit_transform(ge_kidney_cancer_data_correct_labels[gene_cols])" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>columns_for_one_hot = ['project_id', 'Sample_Type']
ge_kidney_cancer_data_correct_labels['project_id_orig'] =ge_kidney_cancer_data_correct_labels['project_id']
ge_kidney_cancer_data_correct_labels['Sample_Type_orig'] = ge_kidney_cancer_data_correct_labels['Sample_Type']
ml_df = pd.get_dummies(ge_kidney_cancer_data_correct_labels, columns=['project_id', 'Sample_Type'], prefix=['cancer_subtype', 'tumor_subtype'])

</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="columns_for_one_hot = [&#039;project_id&#039;, &#039;Sample_Type&#039;]
ge_kidney_cancer_data_correct_labels[&#039;project_id_orig&#039;] =ge_kidney_cancer_data_correct_labels[&#039;project_id&#039;]
ge_kidney_cancer_data_correct_labels[&#039;Sample_Type_orig&#039;] = ge_kidney_cancer_data_correct_labels[&#039;Sample_Type&#039;]
ml_df = pd.get_dummies(ge_kidney_cancer_data_correct_labels, columns=[&#039;project_id&#039;, &#039;Sample_Type&#039;], prefix=[&#039;cancer_subtype&#039;, &#039;tumor_subtype&#039;])" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="131" class="code-line" dir="auto"><code class="hljs">/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2156470527.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  ge_kidney_cancer_data_correct_labels['project_id_orig'] =ge_kidney_cancer_data_correct_labels['project_id']
/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2156470527.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels['project_id_orig'] =ge_kidney_cancer_data_correct_labels['project_id']
/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2156470527.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  ge_kidney_cancer_data_correct_labels['Sample_Type_orig'] = ge_kidney_cancer_data_correct_labels['Sample_Type']
/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2156470527.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels['Sample_Type_orig'] = ge_kidney_cancer_data_correct_labels['Sample_Type']
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2156470527.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  ge_kidney_cancer_data_correct_labels[&#039;project_id_orig&#039;] =ge_kidney_cancer_data_correct_labels[&#039;project_id&#039;]
/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2156470527.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels[&#039;project_id_orig&#039;] =ge_kidney_cancer_data_correct_labels[&#039;project_id&#039;]
/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2156470527.py:3: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead. To get a de-fragmented frame, use `newframe = frame.copy()`
  ge_kidney_cancer_data_correct_labels[&#039;Sample_Type_orig&#039;] = ge_kidney_cancer_data_correct_labels[&#039;Sample_Type&#039;]
/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2156470527.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels[&#039;Sample_Type_orig&#039;] = ge_kidney_cancer_data_correct_labels[&#039;Sample_Type&#039;]" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><h2 id="heading-ii.-exploratory-analysis">II. Exploratory Analysis</h2>
<h2 id="heading-2.1-check-overall-info-of-datatypes2c-null-values">2.1 Check overall info of datatypes, null values</h2>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ge_kidney_cancer_data_correct_labels.info()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ge_kidney_cancer_data_correct_labels.info()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="158" class="code-line" dir="auto"><code class="hljs">&lt;class 'pandas.core.frame.DataFrame'&gt;
Index: 924 entries, 2 to 1293
Columns: 60673 entries, ENSG00000000003.15 to Sample_Type_orig
dtypes: float64(60660), object(13)
memory usage: 427.7+ MB
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="&lt;class &#039;pandas.core.frame.DataFrame&#039;&gt;
Index: 924 entries, 2 to 1293
Columns: 60673 entries, ENSG00000000003.15 to Sample_Type_orig
dtypes: float64(60660), object(13)
memory usage: 427.7+ MB" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><h2 id="heading-2.2.-check-unique-sample-ids">2.2. Check Unique sample IDs</h2>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>print(f"The total patient ids are {ge_kidney_cancer_data_correct_labels['Case_ID'].count()}, from those the unique ids are {ge_kidney_cancer_data_correct_labels['Case_ID'].value_counts().shape[0]} ")
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="print(f&quot;The total patient ids are {ge_kidney_cancer_data_correct_labels[&#039;Case_ID&#039;].count()}, from those the unique ids are {ge_kidney_cancer_data_correct_labels[&#039;Case_ID&#039;].value_counts().shape[0]} &quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="172" class="code-line" dir="auto"><code class="hljs">The total patient ids are 924, from those the unique ids are 799 
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="The total patient ids are 924, from those the unique ids are 799" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><h2 id="heading-2.3.-check-number-of-data-labels-by-project-type-and-tumor-type">2.3. Check number of data labels by project type and tumor type</h2>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ge_kidney_cancer_data_correct_labels.groupby(['Sample_Type'])['Sample_Type'].count()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ge_kidney_cancer_data_correct_labels.groupby([&#039;Sample_Type&#039;])[&#039;Sample_Type&#039;].count()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="185" class="code-line" dir="auto"><code class="hljs">Sample_Type
Additional - New Primary      2
Primary Tumor               800
Solid Tissue Normal         122
Name: Sample_Type, dtype: int64
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Sample_Type
Additional - New Primary      2
Primary Tumor               800
Solid Tissue Normal         122
Name: Sample_Type, dtype: int64" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax = sns.countplot(x='Sample_Type', data=ge_kidney_cancer_data_correct_labels)
ax.bar_label(ax.containers[0])

</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax = sns.countplot(x=&#039;Sample_Type&#039;, data=ge_kidney_cancer_data_correct_labels)
ax.bar_label(ax.containers[0])" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="203" class="code-line" dir="auto"><code class="hljs">[Text(0, 0, '800'), Text(0, 0, '122'), Text(0, 0, '2')]
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="[Text(0, 0, &#039;800&#039;), Text(0, 0, &#039;122&#039;), Text(0, 0, &#039;2&#039;)]" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_16_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_16_1.png" /></p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ge_kidney_cancer_data_correct_labels.groupby(['project_id'])['project_id'].count()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ge_kidney_cancer_data_correct_labels.groupby([&#039;project_id&#039;])[&#039;project_id&#039;].count()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="221" class="code-line" dir="auto"><code class="hljs">project_id
TCGA-KICH     84
TCGA-KIRC    539
TCGA-KIRP    301
Name: project_id, dtype: int64
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="project_id
TCGA-KICH     84
TCGA-KIRC    539
TCGA-KIRP    301
Name: project_id, dtype: int64" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax = sns.countplot(x='project_id', data=ge_kidney_cancer_data_correct_labels)
ax.bar_label(ax.containers[0])

</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax = sns.countplot(x=&#039;project_id&#039;, data=ge_kidney_cancer_data_correct_labels)
ax.bar_label(ax.containers[0])" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="239" class="code-line" dir="auto"><code class="hljs">[Text(0, 0, '539'), Text(0, 0, '301'), Text(0, 0, '84')]
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="[Text(0, 0, &#039;539&#039;), Text(0, 0, &#039;301&#039;), Text(0, 0, &#039;84&#039;)]" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_18_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_18_1.png" /></p><h2 id="heading-iii.-data-sampling2c-augmentation-and-splitting-into-training2c-testing-and-validation-sets-for-supervised-clustering">III. Data sampling, Augmentation and splitting into training, testing and validation sets for Supervised Clustering</h2>
<p>Given that we have imbalanced datasets for Tumor Types and Kidney Cancer subtypes (project_id), we should consider the following for an <strong>unbiased analysis</strong></p><ol data-line="253" class="code-line" dir="auto">
<li><p><strong>Patient overlap</strong>: Given 924 total patients with 799 unique,</p><div  data-line="255" class="code-line" dir="auto" ></div>
<br>
<ol data-line="257" class="code-line" dir="auto">
<li>We should keep same ID patients in only one set i.e (Training or Validation or Testing)</li>
<li>Preferably, we should have have all overlapping patients in Testing so as to not bias our training</li>
</ol>
</li>
</ol>
<div  data-line="260" class="code-line" dir="auto" ></div>
<br>
<ol start="2" data-line="262" class="code-line" dir="auto">
<li><p><strong>Set Sampling</strong>: For minority classes, we sample atleast X% into our testing first followed by validation and then all the remaining in training.</p></li>
<li><p>Additionally for one of the minority classes (New Primary - we should group it into the same category as Primary)</p></li>
<li><p><strong>Data augmention</strong>: Applying oversampling, undersampling techniques for minority or majority classes</p></li>
</ol>
<h2 id="heading-3.1.-label-correction-of-additional---new-primary-tumor">3.1. Label correction of Additional - New Primary tumor</h2>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ge_kidney_cancer_data_correct_labels['Sample_Type'] = ge_kidney_cancer_data_correct_labels['Sample_Type'].apply(lambda x: 'Primary Tumor' if x == 'Additional - New Primary' else x)
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ge_kidney_cancer_data_correct_labels[&#039;Sample_Type&#039;] = ge_kidney_cancer_data_correct_labels[&#039;Sample_Type&#039;].apply(lambda x: &#039;Primary Tumor&#039; if x == &#039;Additional - New Primary&#039; else x)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="275" class="code-line" dir="auto"><code class="hljs">/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2849659451.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels['Sample_Type'] = ge_kidney_cancer_data_correct_labels['Sample_Type'].apply(lambda x: 'Primary Tumor' if x == 'Additional - New Primary' else x)
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="/var/folders/ng/bwk7d4ds7wz95l011dbvtc9r0000gn/T/ipykernel_2912/2849659451.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  ge_kidney_cancer_data_correct_labels[&#039;Sample_Type&#039;] = ge_kidney_cancer_data_correct_labels[&#039;Sample_Type&#039;].apply(lambda x: &#039;Primary Tumor&#039; if x == &#039;Additional - New Primary&#039; else x)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ge_kidney_cancer_data_correct_labels.groupby(['Sample_Type'])['Sample_Type'].count()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ge_kidney_cancer_data_correct_labels.groupby([&#039;Sample_Type&#039;])[&#039;Sample_Type&#039;].count()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="291" class="code-line" dir="auto"><code class="hljs">Sample_Type
Primary Tumor          802
Solid Tissue Normal    122
Name: Sample_Type, dtype: int64
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Sample_Type
Primary Tumor          802
Solid Tissue Normal    122
Name: Sample_Type, dtype: int64" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax = sns.countplot(x='Sample_Type', data=ge_kidney_cancer_data_correct_labels)
ax.bar_label(ax.containers[0])

</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax = sns.countplot(x=&#039;Sample_Type&#039;, data=ge_kidney_cancer_data_correct_labels)
ax.bar_label(ax.containers[0])" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="308" class="code-line" dir="auto"><code class="hljs">[Text(0, 0, '802'), Text(0, 0, '122')]
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="[Text(0, 0, &#039;802&#039;), Text(0, 0, &#039;122&#039;)]" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_23_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_23_1.png" /></p><h2 id="heading-3.2-separation-of-training2c-validation-and-testing-data-for-unbiased-analysis">3.2 Separation of training, validation and testing data for unbiased analysis</h2>
<p>Strategy</p><ol data-line="321" class="code-line" dir="auto">
<li>Find pairs of all duplicate sample IDs and randomly sample pairs 70% in training, 20% in validation and 10% in testing (with seed)</li>
<li>Out of the remaining samples For the 2 lowest groups KICH (10% - 84 samples) and Solid tissue(13% - 122 samples) put more than 25% samples into testing set first
<ol data-line="323" class="code-line" dir="auto">
<li>Take union of 21 samples from KICH (25%) and 30 samples from Solid tissue (25%)  and put all those samples into testing set first</li>
<li>Hold out this set till model training has occurred</li>
<li>Out of the remaining samples, again put more than 25% samples from minority classes into validation</li>
<li>Put all the remaining into training</li>
</ol>
</li>
</ol>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>from sklearn.model_selection import train_test_split

</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="from sklearn.model_selection import train_test_split" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>y_true = ml_df.iloc[:, -6:].drop(['tumor_subtype_Additional - New Primary'], axis=1).astype(int).to_numpy()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="y_true = ml_df.iloc[:, -6:].drop([&#039;tumor_subtype_Additional - New Primary&#039;], axis=1).astype(int).to_numpy()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre># Count of each unique value
value_counts = ml_df['Case_ID'].value_counts()

# list of duplicate ids
duplicate_ids = value_counts[value_counts > 1].reset_index()['Case_ID'].unique()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="# Count of each unique value
value_counts = ml_df[&#039;Case_ID&#039;].value_counts()

# list of duplicate ids
duplicate_ids = value_counts[value_counts &gt; 1].reset_index()[&#039;Case_ID&#039;].unique()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><h2 id="heading-3.2.1-separating-patient-overlap-samples">3.2.1 Separating patient overlap samples</h2>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre># First split: Separate out a test set
train_val_ids, test_overlap_ids = train_test_split(duplicate_ids, test_size=0.1, random_state=42)

# Second split: Separate the remaining data into training and validation sets
train_overlap_ids, val_overlap_ids = train_test_split(train_val_ids, test_size=0.2, random_state=42) # 0.25 x 0.8 = 0.2

print(f"Number of Patient overlap Training IDs: {len(train_overlap_ids)}")
print(f"Number of Patient overlap Validation IDs: {len(val_overlap_ids)}")
print(f"Number of Patient overlap Testing IDs: {len(test_overlap_ids)}")
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="# First split: Separate out a test set
train_val_ids, test_overlap_ids = train_test_split(duplicate_ids, test_size=0.1, random_state=42)

# Second split: Separate the remaining data into training and validation sets
train_overlap_ids, val_overlap_ids = train_test_split(train_val_ids, test_size=0.2, random_state=42) # 0.25 x 0.8 = 0.2

print(f&quot;Number of Patient overlap Training IDs: {len(train_overlap_ids)}&quot;)
print(f&quot;Number of Patient overlap Validation IDs: {len(val_overlap_ids)}&quot;)
print(f&quot;Number of Patient overlap Testing IDs: {len(test_overlap_ids)}&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="364" class="code-line" dir="auto"><code class="hljs">Number of Patient overlap Training IDs: 87
Number of Patient overlap Validation IDs: 22
Number of Patient overlap Testing IDs: 13
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Number of Patient overlap Training IDs: 87
Number of Patient overlap Validation IDs: 22
Number of Patient overlap Testing IDs: 13" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><h2 id="heading-3.2.2-set-sampling">3.2.2 Set sampling</h2>
<p>Shows that majority of duplicate samples came from the Solid tissue class</p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ml_df_without_overlap = ml_df[~ml_df['Case_ID'].isin(duplicate_ids)].reset_index(drop=True)
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ml_df_without_overlap = ml_df[~ml_df[&#039;Case_ID&#039;].isin(duplicate_ids)].reset_index(drop=True)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax1 = sns.countplot(x='Sample_Type', data=ge_kidney_cancer_data_correct_labels)
ax1.bar_label(ax1.containers[0])
ax1.set_title("Count of samples originally across tumors")
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax1 = sns.countplot(x=&#039;Sample_Type&#039;, data=ge_kidney_cancer_data_correct_labels)
ax1.bar_label(ax1.containers[0])
ax1.set_title(&quot;Count of samples originally across tumors&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="388" class="code-line" dir="auto"><code class="hljs">Text(0.5, 1.0, 'Count of samples originally across tumors')
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Text(0.5, 1.0, &#039;Count of samples originally across tumors&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_32_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_32_1.png" /></p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax = sns.countplot(x='Sample_Type_orig', data=ml_df_without_overlap )
ax.bar_label(ax.containers[0])
ax.set_title("Count of samples after separating patient overlap samples")

</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax = sns.countplot(x=&#039;Sample_Type_orig&#039;, data=ml_df_without_overlap )
ax.bar_label(ax.containers[0])
ax.set_title(&quot;Count of samples after separating patient overlap samples&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="409" class="code-line" dir="auto"><code class="hljs">Text(0.5, 1.0, 'Count of samples after separating patient overlap samples')
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Text(0.5, 1.0, &#039;Count of samples after separating patient overlap samples&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_33_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_33_1.png" /></p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax = sns.countplot(x='project_id', data=ge_kidney_cancer_data_correct_labels)
ax.bar_label(ax.containers[0])
ax.set_title("Samples Count in Kidney cancer(KC) subtypes Originally")
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax = sns.countplot(x=&#039;project_id&#039;, data=ge_kidney_cancer_data_correct_labels)
ax.bar_label(ax.containers[0])
ax.set_title(&quot;Samples Count in Kidney cancer(KC) subtypes Originally&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="429" class="code-line" dir="auto"><code class="hljs">Text(0.5, 1.0, 'Samples Count in Kidney cancer(KC) subtypes Originally')
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Text(0.5, 1.0, &#039;Samples Count in Kidney cancer(KC) subtypes Originally&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_34_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_34_1.png" /></p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax = sns.countplot(x='project_id_orig', data=ml_df_without_overlap)
ax.bar_label(ax.containers[0])
ax.set_title("Samples count in KC subtypes after separating patient overlap")
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax = sns.countplot(x=&#039;project_id_orig&#039;, data=ml_df_without_overlap)
ax.bar_label(ax.containers[0])
ax.set_title(&quot;Samples count in KC subtypes after separating patient overlap&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="449" class="code-line" dir="auto"><code class="hljs">Text(0.5, 1.0, 'Samples count in KC subtypes after separating patient overlap')
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Text(0.5, 1.0, &#039;Samples count in KC subtypes after separating patient overlap&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_35_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_35_1.png" /></p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>all_minority_samples1 = ml_df_without_overlap[(ml_df_without_overlap['Sample_Type_orig'] == 'Solid Tissue Normal')
| (ml_df_without_overlap['project_id_orig'] == 'TCGA-KICH')]['Case_ID'].to_numpy()

# First split: Separate out a test set with 50% 
train_val_set_sampl_ids1, test_set_samp_ids1 = train_test_split(all_minority_samples1, test_size=0.5, random_state=42)

# Second split: Separate the remaining data into training and validation sets
train_set_samp_ids1, val_set_samp_ids1 = train_test_split(train_val_set_sampl_ids1, test_size=0.5, random_state=42) 

print(f"Number of Minority set sampled Training IDs: {len(train_set_samp_ids1)}")
print(f"Number of Minority set sampled  Validation IDs: {len(val_set_samp_ids1)}")
print(f"Number of Minority set sampled  Testing IDs: {len(test_set_samp_ids1)}")
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="all_minority_samples1 = ml_df_without_overlap[(ml_df_without_overlap[&#039;Sample_Type_orig&#039;] == &#039;Solid Tissue Normal&#039;)
| (ml_df_without_overlap[&#039;project_id_orig&#039;] == &#039;TCGA-KICH&#039;)][&#039;Case_ID&#039;].to_numpy()

# First split: Separate out a test set with 50% 
train_val_set_sampl_ids1, test_set_samp_ids1 = train_test_split(all_minority_samples1, test_size=0.5, random_state=42)

# Second split: Separate the remaining data into training and validation sets
train_set_samp_ids1, val_set_samp_ids1 = train_test_split(train_val_set_sampl_ids1, test_size=0.5, random_state=42) 

print(f&quot;Number of Minority set sampled Training IDs: {len(train_set_samp_ids1)}&quot;)
print(f&quot;Number of Minority set sampled  Validation IDs: {len(val_set_samp_ids1)}&quot;)
print(f&quot;Number of Minority set sampled  Testing IDs: {len(test_set_samp_ids1)}&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="475" class="code-line" dir="auto"><code class="hljs">Number of Minority set sampled Training IDs: 10
Number of Minority set sampled  Validation IDs: 11
Number of Minority set sampled  Testing IDs: 21
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Number of Minority set sampled Training IDs: 10
Number of Minority set sampled  Validation IDs: 11
Number of Minority set sampled  Testing IDs: 21" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>rem_majority_samples = ml_df_without_overlap[~ml_df_without_overlap['Case_ID'].isin(all_minority_samples1)]['Case_ID'].to_numpy()

# First split: Separate out a test set
train_val_major_ids1, test_major_ids1 = train_test_split(rem_majority_samples, test_size=0.1, random_state=42)

# Second split: Separate the remaining data into training and validation sets
train_major_ids1, val_major_ids1 = train_test_split(train_val_major_ids1, test_size=0.2, random_state=42) # 0.25 x 0.8 = 0.2

print(f"Number of majority classes Training IDs: {len(train_major_ids1)}")
print(f"Number of majority classes Validation IDs: {len(val_major_ids1)}")
print(f"Number of majority classes Testing IDs: {len(test_major_ids1)}")
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="rem_majority_samples = ml_df_without_overlap[~ml_df_without_overlap[&#039;Case_ID&#039;].isin(all_minority_samples1)][&#039;Case_ID&#039;].to_numpy()

# First split: Separate out a test set
train_val_major_ids1, test_major_ids1 = train_test_split(rem_majority_samples, test_size=0.1, random_state=42)

# Second split: Separate the remaining data into training and validation sets
train_major_ids1, val_major_ids1 = train_test_split(train_val_major_ids1, test_size=0.2, random_state=42) # 0.25 x 0.8 = 0.2

print(f&quot;Number of majority classes Training IDs: {len(train_major_ids1)}&quot;)
print(f&quot;Number of majority classes Validation IDs: {len(val_major_ids1)}&quot;)
print(f&quot;Number of majority classes Testing IDs: {len(test_major_ids1)}&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="495" class="code-line" dir="auto"><code class="hljs">Number of majority classes Training IDs: 456
Number of majority classes Validation IDs: 115
Number of majority classes Testing IDs: 64
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Number of majority classes Training IDs: 456
Number of majority classes Validation IDs: 115
Number of majority classes Testing IDs: 64" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>train_ids_df = pd.DataFrame({'Case_ID': list(train_overlap_ids) + list(train_major_ids1) + list(train_set_samp_ids1)}).assign(label='training')
val_ids_df = pd.DataFrame({'Case_ID': list(val_overlap_ids) + list(val_major_ids1) + list(val_set_samp_ids1)}).assign(label='validation')
test_ids_df = pd.DataFrame({'Case_ID': list(test_overlap_ids) + list(test_major_ids1) + list(test_set_samp_ids1)}).assign(label='testing') 
train_val_test_ids_df = pd.concat([train_ids_df, val_ids_df, test_ids_df], axis=0)
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="train_ids_df = pd.DataFrame({&#039;Case_ID&#039;: list(train_overlap_ids) + list(train_major_ids1) + list(train_set_samp_ids1)}).assign(label=&#039;training&#039;)
val_ids_df = pd.DataFrame({&#039;Case_ID&#039;: list(val_overlap_ids) + list(val_major_ids1) + list(val_set_samp_ids1)}).assign(label=&#039;validation&#039;)
test_ids_df = pd.DataFrame({&#039;Case_ID&#039;: list(test_overlap_ids) + list(test_major_ids1) + list(test_set_samp_ids1)}).assign(label=&#039;testing&#039;) 
train_val_test_ids_df = pd.concat([train_ids_df, val_ids_df, test_ids_df], axis=0)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>train_val_test_ids_df = pd.merge(train_val_test_ids_df, ml_df[['Case_ID', 'project_id_orig', 'Sample_Type_orig']],
left_on='Case_ID',
right_on='Case_ID')
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="train_val_test_ids_df = pd.merge(train_val_test_ids_df, ml_df[[&#039;Case_ID&#039;, &#039;project_id_orig&#039;, &#039;Sample_Type_orig&#039;]],
left_on=&#039;Case_ID&#039;,
right_on=&#039;Case_ID&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>train_val_test_ids_df 
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="train_val_test_ids_df" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div  data-line="523" class="code-line" dir="auto" ></div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="529" class="code-line" dir="auto"><code class="hljs">.dataframe tbody tr th {
    vertical-align: top;
}

.dataframe thead th {
    text-align: right;
}
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value=".dataframe tbody tr th {
    vertical-align: top;
}

.dataframe thead th {
    text-align: right;
}" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div  data-line="536" class="code-line" dir="auto" ></div>
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Case_ID</th>
      <th>label</th>
      <th>project_id_orig</th>
      <th>Sample_Type_orig</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>TCGA-CW-5585</td>
      <td>training</td>
      <td>TCGA-KIRC</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>1</th>
      <td>TCGA-CW-5585</td>
      <td>training</td>
      <td>TCGA-KIRC</td>
      <td>Solid Tissue Normal</td>
    </tr>
    <tr>
      <th>2</th>
      <td>TCGA-B0-5706</td>
      <td>training</td>
      <td>TCGA-KIRC</td>
      <td>Solid Tissue Normal</td>
    </tr>
    <tr>
      <th>3</th>
      <td>TCGA-B0-5706</td>
      <td>training</td>
      <td>TCGA-KIRC</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>4</th>
      <td>TCGA-DZ-6132</td>
      <td>training</td>
      <td>TCGA-KIRP</td>
      <td>Solid Tissue Normal</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>919</th>
      <td>TCGA-KN-8418</td>
      <td>testing</td>
      <td>TCGA-KICH</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>920</th>
      <td>TCGA-KO-8406</td>
      <td>testing</td>
      <td>TCGA-KICH</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>921</th>
      <td>TCGA-KO-8414</td>
      <td>testing</td>
      <td>TCGA-KICH</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>922</th>
      <td>TCGA-KL-8325</td>
      <td>testing</td>
      <td>TCGA-KICH</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>923</th>
      <td>TCGA-KM-8438</td>
      <td>testing</td>
      <td>TCGA-KICH</td>
      <td>Primary Tumor</td>
    </tr>
  </tbody>
</table>
<p>924 rows × 4 columns</p>
</div>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>train_val_test_ids_df[['label', 'project_id_orig']].value_counts()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="train_val_test_ids_df[[&#039;label&#039;, &#039;project_id_orig&#039;]].value_counts()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="640" class="code-line" dir="auto"><code class="hljs">label       project_id_orig
training    TCGA-KIRC          394
            TCGA-KIRP          215
validation  TCGA-KIRC           96
            TCGA-KIRP           51
testing     TCGA-KIRC           49
            TCGA-KIRP           35
training    TCGA-KICH           34
testing     TCGA-KICH           27
validation  TCGA-KICH           23
Name: count, dtype: int64
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="label       project_id_orig
training    TCGA-KIRC          394
            TCGA-KIRP          215
validation  TCGA-KIRC           96
            TCGA-KIRP           51
testing     TCGA-KIRC           49
            TCGA-KIRP           35
training    TCGA-KICH           34
testing     TCGA-KICH           27
validation  TCGA-KICH           23
Name: count, dtype: int64" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>train_val_test_ids_df[['label', 'Sample_Type_orig']].value_counts()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="train_val_test_ids_df[[&#039;label&#039;, &#039;Sample_Type_orig&#039;]].value_counts()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="662" class="code-line" dir="auto"><code class="hljs">label       Sample_Type_orig        
training    Primary Tumor               556
validation  Primary Tumor               148
testing     Primary Tumor                96
training    Solid Tissue Normal          86
validation  Solid Tissue Normal          21
testing     Solid Tissue Normal          15
training    Additional - New Primary      1
validation  Additional - New Primary      1
Name: count, dtype: int64
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="label       Sample_Type_orig        
training    Primary Tumor               556
validation  Primary Tumor               148
testing     Primary Tumor                96
training    Solid Tissue Normal          86
validation  Solid Tissue Normal          21
testing     Solid Tissue Normal          15
training    Additional - New Primary      1
validation  Additional - New Primary      1
Name: count, dtype: int64" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>print(f"Final train-validation-test split dataframe has {train_val_test_ids_df.shape[0]} unique samples")
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="print(f&quot;Final train-validation-test split dataframe has {train_val_test_ids_df.shape[0]} unique samples&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="680" class="code-line" dir="auto"><code class="hljs">Final train-validation-test split dataframe has 924 unique samples
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Final train-validation-test split dataframe has 924 unique samples" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><h2 id="heading-iv.-supervised-classification-using-optimized-umap-embedddings-and-clustering">IV. Supervised Classification using optimized UMap embedddings and clustering</h2>
<p>Using training set samples run bayesian optimizer to find the best parameters for 1) Umap (n_components, n_neighbors) 2) HdbScan (min_cluster_size)</p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>clustering_optimizer = OmicsUtils.DimRedMappers.clusterer.ClusteringOptimizer(data=ge_kidney_cancer_data_correct_labels[gene_cols],
                                                                              )
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="clustering_optimizer = OmicsUtils.DimRedMappers.clusterer.ClusteringOptimizer(data=ge_kidney_cancer_data_correct_labels[gene_cols],
                                                                              )" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ge_kidney_cancer_data_correct_labels[gene_cols].to_numpy()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ge_kidney_cancer_data_correct_labels[gene_cols].to_numpy()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="701" class="code-line" dir="auto"><code class="hljs">array([[9.35477035e-01, 1.23067244e-01, 1.36078458e+00, ...,
        0.00000000e+00, 1.13589537e-02, 1.29818744e-01],
       [1.18591878e+00, 1.22465200e-02, 1.41848193e+00, ...,
        0.00000000e+00, 9.54398406e-04, 1.58332332e-01],
       [1.22948986e+00, 4.95363922e-01, 1.27373027e+00, ...,
        0.00000000e+00, 3.50361474e-03, 1.58625081e-02],
       ...,
       [1.15416521e+00, 4.16457433e-01, 1.25601981e+00, ...,
        0.00000000e+00, 3.93420617e-03, 3.61096671e-02],
       [9.14723857e-01, 5.86157970e-02, 1.51866263e+00, ...,
        0.00000000e+00, 7.15010537e-03, 3.39411687e-01],
       [1.32102645e+00, 8.96578816e-02, 1.28390927e+00, ...,
        0.00000000e+00, 2.03640226e-03, 7.48164406e-02]])
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="array([[9.35477035e-01, 1.23067244e-01, 1.36078458e+00, ...,
        0.00000000e+00, 1.13589537e-02, 1.29818744e-01],
       [1.18591878e+00, 1.22465200e-02, 1.41848193e+00, ...,
        0.00000000e+00, 9.54398406e-04, 1.58332332e-01],
       [1.22948986e+00, 4.95363922e-01, 1.27373027e+00, ...,
        0.00000000e+00, 3.50361474e-03, 1.58625081e-02],
       ...,
       [1.15416521e+00, 4.16457433e-01, 1.25601981e+00, ...,
        0.00000000e+00, 3.93420617e-03, 3.61096671e-02],
       [9.14723857e-01, 5.86157970e-02, 1.51866263e+00, ...,
        0.00000000e+00, 7.15010537e-03, 3.39411687e-01],
       [1.32102645e+00, 8.96578816e-02, 1.28390927e+00, ...,
        0.00000000e+00, 2.03640226e-03, 7.48164406e-02]])" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>X = ge_kidney_cancer_data_correct_labels[gene_cols]
label_lower=30
label_upper=100
max_evals = 100
best_param, best_clusters, trials = clustering_optimizer.bayesian_search(embeddings=ge_kidney_cancer_data_correct_labels[gene_cols].to_numpy(), 
                                     label_lower=label_lower,
                                     label_upper=label_upper,
                                     max_evals=max_evals)
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="X = ge_kidney_cancer_data_correct_labels[gene_cols]
label_lower=30
label_upper=100
max_evals = 100
best_param, best_clusters, trials = clustering_optimizer.bayesian_search(embeddings=ge_kidney_cancer_data_correct_labels[gene_cols].to_numpy(), 
                                     label_lower=label_lower,
                                     label_upper=label_upper,
                                     max_evals=max_evals)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="729" class="code-line" dir="auto"><code class="hljs">  0%|          | 0/100 [00:00&lt;?, ?trial/s, best loss=?]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  1%|          | 1/100 [00:27&lt;44:58, 27.26s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  2%|▏         | 2/100 [00:52&lt;42:42, 26.14s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  3%|▎         | 3/100 [01:19&lt;42:42, 26.42s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  4%|▍         | 4/100 [01:45&lt;41:55, 26.20s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  5%|▌         | 5/100 [02:12&lt;41:53, 26.46s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  6%|▌         | 6/100 [02:37&lt;41:03, 26.21s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  7%|▋         | 7/100 [03:03&lt;40:24, 26.07s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  8%|▊         | 8/100 [03:28&lt;39:36, 25.83s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  9%|▉         | 9/100 [03:54&lt;38:48, 25.59s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 10%|█         | 10/100 [04:18&lt;38:03, 25.37s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 11%|█         | 11/100 [04:44&lt;37:41, 25.41s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 12%|█▏        | 12/100 [05:09&lt;37:11, 25.36s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 13%|█▎        | 13/100 [05:35&lt;37:02, 25.55s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 14%|█▍        | 14/100 [06:01&lt;36:47, 25.67s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 15%|█▌        | 15/100 [06:27&lt;36:26, 25.72s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 16%|█▌        | 16/100 [06:52&lt;35:53, 25.64s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 17%|█▋        | 17/100 [07:18&lt;35:29, 25.66s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 18%|█▊        | 18/100 [07:44&lt;35:04, 25.66s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 19%|█▉        | 19/100 [08:10&lt;34:42, 25.71s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 20%|██        | 20/100 [08:35&lt;34:19, 25.75s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 21%|██        | 21/100 [09:01&lt;33:54, 25.75s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 22%|██▏       | 22/100 [09:27&lt;33:31, 25.79s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 23%|██▎       | 23/100 [09:52&lt;32:57, 25.68s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 24%|██▍       | 24/100 [10:18&lt;32:27, 25.62s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 25%|██▌       | 25/100 [10:45&lt;32:35, 26.07s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 26%|██▌       | 26/100 [11:11&lt;32:09, 26.08s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 27%|██▋       | 27/100 [11:36&lt;31:23, 25.80s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 28%|██▊       | 28/100 [12:01&lt;30:36, 25.51s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 29%|██▉       | 29/100 [12:26&lt;29:56, 25.31s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 30%|███       | 30/100 [12:52&lt;29:45, 25.51s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 31%|███       | 31/100 [13:18&lt;29:31, 25.67s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 32%|███▏      | 32/100 [13:44&lt;29:21, 25.90s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 33%|███▎      | 33/100 [14:18&lt;31:21, 28.08s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 34%|███▍      | 34/100 [14:53&lt;33:22, 30.34s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 35%|███▌      | 35/100 [15:25&lt;33:26, 30.88s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 36%|███▌      | 36/100 [15:56&lt;32:54, 30.86s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 37%|███▋      | 37/100 [16:25&lt;31:54, 30.39s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 38%|███▊      | 38/100 [16:54&lt;30:44, 29.75s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 39%|███▉      | 39/100 [17:22&lt;29:42, 29.22s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 40%|████      | 40/100 [17:49&lt;28:45, 28.76s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 41%|████      | 41/100 [18:17&lt;27:55, 28.39s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 42%|████▏     | 42/100 [18:44&lt;27:07, 28.07s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 43%|████▎     | 43/100 [19:10&lt;25:52, 27.24s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 44%|████▍     | 44/100 [19:37&lt;25:26, 27.25s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 45%|████▌     | 45/100 [20:06&lt;25:27, 27.77s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 46%|████▌     | 46/100 [20:35&lt;25:21, 28.18s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 47%|████▋     | 47/100 [21:04&lt;25:08, 28.47s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 48%|████▊     | 48/100 [21:33&lt;24:40, 28.48s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 49%|████▉     | 49/100 [22:01&lt;24:15, 28.54s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 50%|█████     | 50/100 [22:30&lt;23:48, 28.56s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 51%|█████     | 51/100 [22:58&lt;23:13, 28.44s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 52%|█████▏    | 52/100 [23:26&lt;22:42, 28.39s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 53%|█████▎    | 53/100 [23:54&lt;22:05, 28.20s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 54%|█████▍    | 54/100 [24:22&lt;21:33, 28.13s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 55%|█████▌    | 55/100 [24:50&lt;21:04, 28.11s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 56%|█████▌    | 56/100 [25:18&lt;20:37, 28.13s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 57%|█████▋    | 57/100 [25:46&lt;20:07, 28.09s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 58%|█████▊    | 58/100 [26:15&lt;19:41, 28.14s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 59%|█████▉    | 59/100 [26:43&lt;19:11, 28.09s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 60%|██████    | 60/100 [27:11&lt;18:46, 28.17s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 61%|██████    | 61/100 [27:39&lt;18:16, 28.13s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 62%|██████▏   | 62/100 [28:07&lt;17:50, 28.18s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 63%|██████▎   | 63/100 [28:36&lt;17:26, 28.28s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 64%|██████▍   | 64/100 [29:04&lt;17:02, 28.41s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 65%|██████▌   | 65/100 [29:33&lt;16:34, 28.42s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 66%|██████▌   | 66/100 [30:01&lt;16:07, 28.44s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 67%|██████▋   | 67/100 [30:30&lt;15:45, 28.64s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 68%|██████▊   | 68/100 [31:00&lt;15:28, 29.03s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 69%|██████▉   | 69/100 [31:30&lt;15:07, 29.28s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 70%|███████   | 70/100 [31:59&lt;14:35, 29.19s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 71%|███████   | 71/100 [32:28&lt;14:02, 29.06s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 72%|███████▏  | 72/100 [32:56&lt;13:26, 28.81s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 73%|███████▎  | 73/100 [33:25&lt;12:53, 28.65s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 74%|███████▍  | 74/100 [33:53&lt;12:23, 28.58s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 75%|███████▌  | 75/100 [34:21&lt;11:49, 28.40s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 76%|███████▌  | 76/100 [34:49&lt;11:22, 28.43s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 77%|███████▋  | 77/100 [35:17&lt;10:50, 28.26s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 78%|███████▊  | 78/100 [35:45&lt;10:18, 28.13s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 79%|███████▉  | 79/100 [36:13&lt;09:48, 28.00s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 80%|████████  | 80/100 [36:40&lt;09:17, 27.90s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 81%|████████  | 81/100 [37:08&lt;08:50, 27.93s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 82%|████████▏ | 82/100 [37:36&lt;08:22, 27.93s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 83%|████████▎ | 83/100 [38:04&lt;07:53, 27.88s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 84%|████████▍ | 84/100 [38:33&lt;07:29, 28.08s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 85%|████████▌ | 85/100 [39:01&lt;07:00, 28.01s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 86%|████████▌ | 86/100 [39:28&lt;06:31, 27.98s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 87%|████████▋ | 87/100 [39:56&lt;06:02, 27.91s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 88%|████████▊ | 88/100 [40:24&lt;05:35, 27.96s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 89%|████████▉ | 89/100 [40:52&lt;05:06, 27.87s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 90%|█████████ | 90/100 [41:20&lt;04:40, 28.07s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 91%|█████████ | 91/100 [41:49&lt;04:13, 28.16s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 92%|█████████▏| 92/100 [42:17&lt;03:45, 28.15s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 93%|█████████▎| 93/100 [42:45&lt;03:16, 28.12s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 94%|█████████▍| 94/100 [43:13&lt;02:47, 27.98s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 95%|█████████▌| 95/100 [43:43&lt;02:24, 28.81s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 96%|█████████▌| 96/100 [44:13&lt;01:56, 29.09s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 97%|█████████▋| 97/100 [44:42&lt;01:27, 29.16s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 98%|█████████▊| 98/100 [45:12&lt;00:58, 29.15s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 99%|█████████▉| 99/100 [45:41&lt;00:29, 29.12s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



100%|██████████| 100/100 [46:12&lt;00:00, 27.72s/trial, best loss: 0.05]
best:
{'min_cluster_size': 7, 'n_components': 6, 'n_neighbors': 5, 'random_state': 42}
label count: 6


/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="0%|          | 0/100 [00:00&lt;?, ?trial/s, best loss=?]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  1%|          | 1/100 [00:27&lt;44:58, 27.26s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  2%|▏         | 2/100 [00:52&lt;42:42, 26.14s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  3%|▎         | 3/100 [01:19&lt;42:42, 26.42s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  4%|▍         | 4/100 [01:45&lt;41:55, 26.20s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  5%|▌         | 5/100 [02:12&lt;41:53, 26.46s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  6%|▌         | 6/100 [02:37&lt;41:03, 26.21s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  7%|▋         | 7/100 [03:03&lt;40:24, 26.07s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  8%|▊         | 8/100 [03:28&lt;39:36, 25.83s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



  9%|▉         | 9/100 [03:54&lt;38:48, 25.59s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 10%|█         | 10/100 [04:18&lt;38:03, 25.37s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 11%|█         | 11/100 [04:44&lt;37:41, 25.41s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 12%|█▏        | 12/100 [05:09&lt;37:11, 25.36s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 13%|█▎        | 13/100 [05:35&lt;37:02, 25.55s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 14%|█▍        | 14/100 [06:01&lt;36:47, 25.67s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 15%|█▌        | 15/100 [06:27&lt;36:26, 25.72s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 16%|█▌        | 16/100 [06:52&lt;35:53, 25.64s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 17%|█▋        | 17/100 [07:18&lt;35:29, 25.66s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 18%|█▊        | 18/100 [07:44&lt;35:04, 25.66s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 19%|█▉        | 19/100 [08:10&lt;34:42, 25.71s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 20%|██        | 20/100 [08:35&lt;34:19, 25.75s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 21%|██        | 21/100 [09:01&lt;33:54, 25.75s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 22%|██▏       | 22/100 [09:27&lt;33:31, 25.79s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 23%|██▎       | 23/100 [09:52&lt;32:57, 25.68s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 24%|██▍       | 24/100 [10:18&lt;32:27, 25.62s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 25%|██▌       | 25/100 [10:45&lt;32:35, 26.07s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 26%|██▌       | 26/100 [11:11&lt;32:09, 26.08s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 27%|██▋       | 27/100 [11:36&lt;31:23, 25.80s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 28%|██▊       | 28/100 [12:01&lt;30:36, 25.51s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 29%|██▉       | 29/100 [12:26&lt;29:56, 25.31s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 30%|███       | 30/100 [12:52&lt;29:45, 25.51s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 31%|███       | 31/100 [13:18&lt;29:31, 25.67s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 32%|███▏      | 32/100 [13:44&lt;29:21, 25.90s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 33%|███▎      | 33/100 [14:18&lt;31:21, 28.08s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 34%|███▍      | 34/100 [14:53&lt;33:22, 30.34s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 35%|███▌      | 35/100 [15:25&lt;33:26, 30.88s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 36%|███▌      | 36/100 [15:56&lt;32:54, 30.86s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 37%|███▋      | 37/100 [16:25&lt;31:54, 30.39s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 38%|███▊      | 38/100 [16:54&lt;30:44, 29.75s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 39%|███▉      | 39/100 [17:22&lt;29:42, 29.22s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 40%|████      | 40/100 [17:49&lt;28:45, 28.76s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 41%|████      | 41/100 [18:17&lt;27:55, 28.39s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 42%|████▏     | 42/100 [18:44&lt;27:07, 28.07s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 43%|████▎     | 43/100 [19:10&lt;25:52, 27.24s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 44%|████▍     | 44/100 [19:37&lt;25:26, 27.25s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 45%|████▌     | 45/100 [20:06&lt;25:27, 27.77s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 46%|████▌     | 46/100 [20:35&lt;25:21, 28.18s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 47%|████▋     | 47/100 [21:04&lt;25:08, 28.47s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 48%|████▊     | 48/100 [21:33&lt;24:40, 28.48s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 49%|████▉     | 49/100 [22:01&lt;24:15, 28.54s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 50%|█████     | 50/100 [22:30&lt;23:48, 28.56s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 51%|█████     | 51/100 [22:58&lt;23:13, 28.44s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 52%|█████▏    | 52/100 [23:26&lt;22:42, 28.39s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 53%|█████▎    | 53/100 [23:54&lt;22:05, 28.20s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 54%|█████▍    | 54/100 [24:22&lt;21:33, 28.13s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 55%|█████▌    | 55/100 [24:50&lt;21:04, 28.11s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 56%|█████▌    | 56/100 [25:18&lt;20:37, 28.13s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 57%|█████▋    | 57/100 [25:46&lt;20:07, 28.09s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 58%|█████▊    | 58/100 [26:15&lt;19:41, 28.14s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 59%|█████▉    | 59/100 [26:43&lt;19:11, 28.09s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 60%|██████    | 60/100 [27:11&lt;18:46, 28.17s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 61%|██████    | 61/100 [27:39&lt;18:16, 28.13s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 62%|██████▏   | 62/100 [28:07&lt;17:50, 28.18s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 63%|██████▎   | 63/100 [28:36&lt;17:26, 28.28s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 64%|██████▍   | 64/100 [29:04&lt;17:02, 28.41s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 65%|██████▌   | 65/100 [29:33&lt;16:34, 28.42s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 66%|██████▌   | 66/100 [30:01&lt;16:07, 28.44s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 67%|██████▋   | 67/100 [30:30&lt;15:45, 28.64s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 68%|██████▊   | 68/100 [31:00&lt;15:28, 29.03s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 69%|██████▉   | 69/100 [31:30&lt;15:07, 29.28s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 70%|███████   | 70/100 [31:59&lt;14:35, 29.19s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 71%|███████   | 71/100 [32:28&lt;14:02, 29.06s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 72%|███████▏  | 72/100 [32:56&lt;13:26, 28.81s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 73%|███████▎  | 73/100 [33:25&lt;12:53, 28.65s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 74%|███████▍  | 74/100 [33:53&lt;12:23, 28.58s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 75%|███████▌  | 75/100 [34:21&lt;11:49, 28.40s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 76%|███████▌  | 76/100 [34:49&lt;11:22, 28.43s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 77%|███████▋  | 77/100 [35:17&lt;10:50, 28.26s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 78%|███████▊  | 78/100 [35:45&lt;10:18, 28.13s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 79%|███████▉  | 79/100 [36:13&lt;09:48, 28.00s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 80%|████████  | 80/100 [36:40&lt;09:17, 27.90s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 81%|████████  | 81/100 [37:08&lt;08:50, 27.93s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 82%|████████▏ | 82/100 [37:36&lt;08:22, 27.93s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 83%|████████▎ | 83/100 [38:04&lt;07:53, 27.88s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 84%|████████▍ | 84/100 [38:33&lt;07:29, 28.08s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 85%|████████▌ | 85/100 [39:01&lt;07:00, 28.01s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 86%|████████▌ | 86/100 [39:28&lt;06:31, 27.98s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 87%|████████▋ | 87/100 [39:56&lt;06:02, 27.91s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 88%|████████▊ | 88/100 [40:24&lt;05:35, 27.96s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 89%|████████▉ | 89/100 [40:52&lt;05:06, 27.87s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 90%|█████████ | 90/100 [41:20&lt;04:40, 28.07s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 91%|█████████ | 91/100 [41:49&lt;04:13, 28.16s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 92%|█████████▏| 92/100 [42:17&lt;03:45, 28.15s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 93%|█████████▎| 93/100 [42:45&lt;03:16, 28.12s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 94%|█████████▍| 94/100 [43:13&lt;02:47, 27.98s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 95%|█████████▌| 95/100 [43:43&lt;02:24, 28.81s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 96%|█████████▌| 96/100 [44:13&lt;01:56, 29.09s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 97%|█████████▋| 97/100 [44:42&lt;01:27, 29.16s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 98%|█████████▊| 98/100 [45:12&lt;00:58, 29.15s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



 99%|█████████▉| 99/100 [45:41&lt;00:29, 29.12s/trial, best loss: 0.05]

/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)



100%|██████████| 100/100 [46:12&lt;00:00, 27.72s/trial, best loss: 0.05]
best:
{&#039;min_cluster_size&#039;: 7, &#039;n_components&#039;: 6, &#039;n_neighbors&#039;: 5, &#039;random_state&#039;: 42}
label count: 6


/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><h2 id="heading-generate-clusters-using-best-params">Generate clusters using best params</h2>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>clusters, umap_embeddings = clustering_optimizer.generate_clusters(message_embeddings=ge_kidney_cancer_data_correct_labels[gene_cols].to_numpy(),
                        n_neighbors=5,
                        n_components=6,
                        clust_params={"min_cluster_size": 7},   
                        random_state = 42)
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="clusters, umap_embeddings = clustering_optimizer.generate_clusters(message_embeddings=ge_kidney_cancer_data_correct_labels[gene_cols].to_numpy(),
                        n_neighbors=5,
                        n_components=6,
                        clust_params={&quot;min_cluster_size&quot;: 7},   
                        random_state = 42)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="1450" class="code-line" dir="auto"><code class="hljs">/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="/opt/homebrew/anaconda3/envs/umap-env/lib/python3.11/site-packages/umap/umap_.py:1943: UserWarning: n_jobs value -1 overridden to 1 by setting random_state. Use no seed for parallelism.
  warn(f&quot;n_jobs value {self.n_jobs} overridden to 1 by setting random_state. Use no seed for parallelism.&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ge_kidney_cancer_data_correct_labels
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ge_kidney_cancer_data_correct_labels" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div  data-line="1462" class="code-line" dir="auto" ></div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="1468" class="code-line" dir="auto"><code class="hljs">.dataframe tbody tr th {
    vertical-align: top;
}

.dataframe thead th {
    text-align: right;
}
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value=".dataframe tbody tr th {
    vertical-align: top;
}

.dataframe thead th {
    text-align: right;
}" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div  data-line="1475" class="code-line" dir="auto" ></div>
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>ENSG00000000003.15</th>
      <th>ENSG00000000005.6</th>
      <th>ENSG00000000419.13</th>
      <th>ENSG00000000457.14</th>
      <th>ENSG00000000460.17</th>
      <th>ENSG00000000938.13</th>
      <th>ENSG00000000971.16</th>
      <th>ENSG00000001036.14</th>
      <th>ENSG00000001084.13</th>
      <th>ENSG00000001167.14</th>
      <th>...</th>
      <th>File ID</th>
      <th>File Name</th>
      <th>Data Category</th>
      <th>Data Type</th>
      <th>Project ID</th>
      <th>Case ID</th>
      <th>Sample ID</th>
      <th>Sample Type</th>
      <th>case_submitter_id</th>
      <th>project_id</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>2</th>
      <td>0.935477</td>
      <td>0.123067</td>
      <td>1.360785</td>
      <td>0.452966</td>
      <td>0.211307</td>
      <td>1.220691</td>
      <td>1.203055</td>
      <td>1.257009</td>
      <td>0.630936</td>
      <td>0.760030</td>
      <td>...</td>
      <td>2f9b9698-edc6-4049-809c-9733f05f5c2b</td>
      <td>ed920009-6bce-4300-b928-14d0ec474ac6.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KIRC</td>
      <td>TCGA-MM-A563</td>
      <td>TCGA-MM-A563-01A</td>
      <td>Primary Tumor</td>
      <td>TCGA-MM-A563</td>
      <td>TCGA-KIRC</td>
    </tr>
    <tr>
      <th>3</th>
      <td>1.185919</td>
      <td>0.012247</td>
      <td>1.418482</td>
      <td>0.222404</td>
      <td>0.065468</td>
      <td>0.344648</td>
      <td>0.111834</td>
      <td>1.503345</td>
      <td>0.875102</td>
      <td>0.890488</td>
      <td>...</td>
      <td>0a4e4402-13a5-4eea-b70c-b121bcf81156</td>
      <td>6abcbd13-ff11-46d6-9ce0-6204a6658c39.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KIRP</td>
      <td>TCGA-GL-8500</td>
      <td>TCGA-GL-8500-01A</td>
      <td>Primary Tumor</td>
      <td>TCGA-GL-8500</td>
      <td>TCGA-KIRP</td>
    </tr>
    <tr>
      <th>4</th>
      <td>1.229490</td>
      <td>0.495364</td>
      <td>1.273730</td>
      <td>0.385338</td>
      <td>0.082606</td>
      <td>0.446848</td>
      <td>0.738614</td>
      <td>1.363431</td>
      <td>0.786524</td>
      <td>0.742269</td>
      <td>...</td>
      <td>091d9ed1-6127-4b37-b2d5-a463b28e2e9e</td>
      <td>c8d14d90-63a8-4576-ae80-ef8582792d16.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KIRP</td>
      <td>TCGA-BQ-5877</td>
      <td>TCGA-BQ-5877-11A</td>
      <td>Solid Tissue Normal</td>
      <td>TCGA-BQ-5877</td>
      <td>TCGA-KIRP</td>
    </tr>
    <tr>
      <th>5</th>
      <td>1.326889</td>
      <td>0.000000</td>
      <td>1.241932</td>
      <td>0.471732</td>
      <td>0.225774</td>
      <td>0.794509</td>
      <td>1.370754</td>
      <td>1.592809</td>
      <td>0.618038</td>
      <td>1.189510</td>
      <td>...</td>
      <td>dcb89df3-3699-45fb-a636-9bff204ff377</td>
      <td>e47989a4-a2f7-46b7-9244-c91e0ae0ee0f.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KIRP</td>
      <td>TCGA-BQ-5877</td>
      <td>TCGA-BQ-5877-01A</td>
      <td>Primary Tumor</td>
      <td>TCGA-BQ-5877</td>
      <td>TCGA-KIRP</td>
    </tr>
    <tr>
      <th>6</th>
      <td>1.490151</td>
      <td>0.052886</td>
      <td>1.323252</td>
      <td>0.518553</td>
      <td>0.125839</td>
      <td>0.342955</td>
      <td>0.294797</td>
      <td>1.262937</td>
      <td>0.675292</td>
      <td>0.695587</td>
      <td>...</td>
      <td>c405ed6b-d3db-4db2-9686-ff8c0a4089b7</td>
      <td>df99ecaa-143c-4dbf-91c5-8e3a76c60f74.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KICH</td>
      <td>TCGA-KN-8423</td>
      <td>TCGA-KN-8423-01A</td>
      <td>Primary Tumor</td>
      <td>TCGA-KN-8423</td>
      <td>TCGA-KICH</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>1287</th>
      <td>1.236482</td>
      <td>0.123035</td>
      <td>1.469803</td>
      <td>0.588025</td>
      <td>0.263423</td>
      <td>0.967061</td>
      <td>0.726124</td>
      <td>1.423254</td>
      <td>0.807474</td>
      <td>1.037223</td>
      <td>...</td>
      <td>68b2e11e-37c9-4e2b-a532-fa8f91424694</td>
      <td>e045554d-52a6-4e83-8ca1-0cbb75a2a68b.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KIRP</td>
      <td>TCGA-B3-3925</td>
      <td>TCGA-B3-3925-01A</td>
      <td>Primary Tumor</td>
      <td>TCGA-B3-3925</td>
      <td>TCGA-KIRP</td>
    </tr>
    <tr>
      <th>1290</th>
      <td>1.267758</td>
      <td>0.074999</td>
      <td>1.257655</td>
      <td>0.242019</td>
      <td>0.086467</td>
      <td>0.450988</td>
      <td>0.394469</td>
      <td>1.383887</td>
      <td>0.795908</td>
      <td>0.582881</td>
      <td>...</td>
      <td>a4ac446e-4c70-4360-9e88-879641cff075</td>
      <td>dca406a5-9351-41b4-8cc6-40ed702103f6.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KIRP</td>
      <td>TCGA-A4-8518</td>
      <td>TCGA-A4-8518-01A</td>
      <td>Primary Tumor</td>
      <td>TCGA-A4-8518</td>
      <td>TCGA-KIRP</td>
    </tr>
    <tr>
      <th>1291</th>
      <td>1.154165</td>
      <td>0.416457</td>
      <td>1.256020</td>
      <td>0.440074</td>
      <td>0.208173</td>
      <td>0.667322</td>
      <td>1.144951</td>
      <td>1.280795</td>
      <td>0.678964</td>
      <td>0.964684</td>
      <td>...</td>
      <td>8c4bd061-8371-4313-bfdc-1b2213daea12</td>
      <td>d85f1c60-f3f3-4355-9f31-6c0e5d53360b.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KIRC</td>
      <td>TCGA-B8-5165</td>
      <td>TCGA-B8-5165-01A</td>
      <td>Primary Tumor</td>
      <td>TCGA-B8-5165</td>
      <td>TCGA-KIRC</td>
    </tr>
    <tr>
      <th>1292</th>
      <td>0.914724</td>
      <td>0.058616</td>
      <td>1.518663</td>
      <td>0.371622</td>
      <td>0.140225</td>
      <td>0.681187</td>
      <td>1.172004</td>
      <td>1.386880</td>
      <td>0.520064</td>
      <td>0.919235</td>
      <td>...</td>
      <td>ba209e1e-87f1-47a9-a858-d60953ef4528</td>
      <td>28089de8-830d-4774-937f-97d6dd7cfa0a.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KIRC</td>
      <td>TCGA-B0-4821</td>
      <td>TCGA-B0-4821-01A</td>
      <td>Primary Tumor</td>
      <td>TCGA-B0-4821</td>
      <td>TCGA-KIRC</td>
    </tr>
    <tr>
      <th>1293</th>
      <td>1.321026</td>
      <td>0.089658</td>
      <td>1.283909</td>
      <td>0.403275</td>
      <td>0.157880</td>
      <td>0.674815</td>
      <td>1.000130</td>
      <td>1.615682</td>
      <td>0.900159</td>
      <td>0.952187</td>
      <td>...</td>
      <td>4a2da9f0-b75d-466c-b9f0-2e910ee24365</td>
      <td>2cf32f8f-0698-453d-a96b-c66a48617d2d.rna_seq.a...</td>
      <td>Transcriptome Profiling</td>
      <td>Gene Expression Quantification</td>
      <td>TCGA-KIRC</td>
      <td>TCGA-AK-3456</td>
      <td>TCGA-AK-3456-01A</td>
      <td>Primary Tumor</td>
      <td>TCGA-AK-3456</td>
      <td>TCGA-KIRC</td>
    </tr>
  </tbody>
</table>
<p>924 rows × 60671 columns</p>
</div>
<h2 id="heading-create-dataframe-of-labels-and-cluster-labels">Create dataframe of labels and cluster labels</h2>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>embed_x, embed_y = umap_embeddings[:, 0], umap_embeddings[:, 1]
umap_embedding_df =  pd.DataFrame({'embed_x': embed_x,
                                     'embed_y': embed_y}
                                     )

umap_embedding_df['project_labels'] = ge_kidney_cancer_data_correct_labels['project_id'].to_numpy()
umap_embedding_df['cluster_labels'] = clusters.labels_
umap_embedding_df['tumor_type_labels'] = ge_kidney_cancer_data_correct_labels['Sample Type'].to_numpy()
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="embed_x, embed_y = umap_embeddings[:, 0], umap_embeddings[:, 1]
umap_embedding_df =  pd.DataFrame({&#039;embed_x&#039;: embed_x,
                                     &#039;embed_y&#039;: embed_y}
                                     )

umap_embedding_df[&#039;project_labels&#039;] = ge_kidney_cancer_data_correct_labels[&#039;project_id&#039;].to_numpy()
umap_embedding_df[&#039;cluster_labels&#039;] = clusters.labels_
umap_embedding_df[&#039;tumor_type_labels&#039;] = ge_kidney_cancer_data_correct_labels[&#039;Sample Type&#039;].to_numpy()" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>umap_embedding_df
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="umap_embedding_df" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div  data-line="1797" class="code-line" dir="auto" ></div>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="1803" class="code-line" dir="auto"><code class="hljs">.dataframe tbody tr th {
    vertical-align: top;
}

.dataframe thead th {
    text-align: right;
}
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value=".dataframe tbody tr th {
    vertical-align: top;
}

.dataframe thead th {
    text-align: right;
}" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div  data-line="1810" class="code-line" dir="auto" ></div>
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>embed_x</th>
      <th>embed_y</th>
      <th>project_labels</th>
      <th>cluster_labels</th>
      <th>tumor_type_labels</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>10.298291</td>
      <td>6.447575</td>
      <td>TCGA-KIRC</td>
      <td>2</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>1</th>
      <td>9.352697</td>
      <td>4.007873</td>
      <td>TCGA-KIRP</td>
      <td>3</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>2</th>
      <td>1.777693</td>
      <td>0.079027</td>
      <td>TCGA-KIRP</td>
      <td>5</td>
      <td>Solid Tissue Normal</td>
    </tr>
    <tr>
      <th>3</th>
      <td>9.620252</td>
      <td>4.031096</td>
      <td>TCGA-KIRP</td>
      <td>3</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>4</th>
      <td>-1.964766</td>
      <td>11.059839</td>
      <td>TCGA-KICH</td>
      <td>0</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>919</th>
      <td>9.150193</td>
      <td>3.324971</td>
      <td>TCGA-KIRP</td>
      <td>3</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>920</th>
      <td>10.274561</td>
      <td>3.604016</td>
      <td>TCGA-KIRP</td>
      <td>3</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>921</th>
      <td>10.978981</td>
      <td>6.363336</td>
      <td>TCGA-KIRC</td>
      <td>2</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>922</th>
      <td>9.980220</td>
      <td>6.798731</td>
      <td>TCGA-KIRC</td>
      <td>2</td>
      <td>Primary Tumor</td>
    </tr>
    <tr>
      <th>923</th>
      <td>11.133520</td>
      <td>3.426035</td>
      <td>TCGA-KIRC</td>
      <td>3</td>
      <td>Primary Tumor</td>
    </tr>
  </tbody>
</table>
<p>924 rows × 5 columns</p>
</div>
<h2 id="heading-plot-results">Plot results</h2>
<div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>import matplotlib.pyplot as plt
import seaborn as sns
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="import matplotlib.pyplot as plt
import seaborn as sns" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax3 = sns.scatterplot(data=umap_embedding_df,
x="embed_x", y="embed_y", hue="cluster_labels")
ax3.set_title("HDBSCAN Clustering with bayesian optimized clusters")


</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax3 = sns.scatterplot(data=umap_embedding_df,
x=&quot;embed_x&quot;, y=&quot;embed_y&quot;, hue=&quot;cluster_labels&quot;)
ax3.set_title(&quot;HDBSCAN Clustering with bayesian optimized clusters&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="1938" class="code-line" dir="auto"><code class="hljs">Text(0.5, 1.0, 'HDBSCAN Clustering with bayesian optimized clusters')
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Text(0.5, 1.0, &#039;HDBSCAN Clustering with bayesian optimized clusters&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_56_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_56_1.png" /></p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax2 = sns.scatterplot(data = 
umap_embedding_df, x='embed_x', y='embed_y', hue='project_labels')
ax2.set_title("HDBSCAN Clustering with ground truth labels")
</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax2 = sns.scatterplot(data = 
umap_embedding_df, x=&#039;embed_x&#039;, y=&#039;embed_y&#039;, hue=&#039;project_labels&#039;)
ax2.set_title(&quot;HDBSCAN Clustering with ground truth labels&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="1958" class="code-line" dir="auto"><code class="hljs">Text(0.5, 1.0, 'HDBSCAN Clustering with ground truth labels')
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Text(0.5, 1.0, &#039;HDBSCAN Clustering with ground truth labels&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_57_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_57_1.png" /></p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>ax1 = sns.scatterplot(data = 
umap_embedding_df, x='embed_x', y='embed_y', hue='tumor_type_labels')
ax1.set_title("HDBSCAN Clustering with ground truth labels")

</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="ax1 = sns.scatterplot(data = 
umap_embedding_df, x=&#039;embed_x&#039;, y=&#039;embed_y&#039;, hue=&#039;tumor_type_labels&#039;)
ax1.set_title(&quot;HDBSCAN Clustering with ground truth labels&quot;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre data-line="1979" class="code-line" dir="auto"><code class="hljs">Text(0.5, 1.0, 'HDBSCAN Clustering with ground truth labels')
</code></pre>

              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="Text(0.5, 1.0, &#039;HDBSCAN Clustering with ground truth labels&#039;)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_58_1.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_58_1.png" /></p><div class="snippet-clipboard-content notranslate position-relative overflow-auto">
              <pre>## combining plots


fig, axes = plt.subplots(3, 1, figsize=(10, 20))
 


sns.scatterplot(ax=axes[0], data=umap_embedding_df, x='embed_x', y='embed_y', hue='tumor_type_labels')
sns.scatterplot(ax=axes[1], data=umap_embedding_df, x='embed_x', y='embed_y', hue='cluster_labels')
sns.scatterplot(ax=axes[2], data=umap_embedding_df, x='embed_x', y='embed_y', hue='project_labels')
fig.suptitle('Umap(Cosine) + HdbScan on Kidney cancer GE data', fontsize = 20)
fig = axes[0].get_figure()
fig.tight_layout()
fig.subplots_adjust(top=0.95)

</pre>
              <div class="zeroclipboard-container position-absolute right-0 top-0">
                <clipboard-copy aria-label="Copy" class="ClipboardButton clipboard-copy m-2 p-0 tooltipped-no-delay"
                  value="## combining plots


fig, axes = plt.subplots(3, 1, figsize=(10, 20))
 


sns.scatterplot(ax=axes[0], data=umap_embedding_df, x=&#039;embed_x&#039;, y=&#039;embed_y&#039;, hue=&#039;tumor_type_labels&#039;)
sns.scatterplot(ax=axes[1], data=umap_embedding_df, x=&#039;embed_x&#039;, y=&#039;embed_y&#039;, hue=&#039;cluster_labels&#039;)
sns.scatterplot(ax=axes[2], data=umap_embedding_df, x=&#039;embed_x&#039;, y=&#039;embed_y&#039;, hue=&#039;project_labels&#039;)
fig.suptitle(&#039;Umap(Cosine) + HdbScan on Kidney cancer GE data&#039;, fontsize = 20)
fig = axes[0].get_figure()
fig.tight_layout()
fig.subplots_adjust(top=0.95)" tabindex="0" role="button">
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-copy clipboard-copy-icon m-2">
                    <path
                      d="M0 6.75C0 5.784.784 5 1.75 5h1.5a.75.75 0 0 1 0 1.5h-1.5a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-1.5a.75.75 0 0 1 1.5 0v1.5A1.75 1.75 0 0 1 9.25 16h-7.5A1.75 1.75 0 0 1 0 14.25Z">
                    </path>
                    <path
                      d="M5 1.75C5 .784 5.784 0 6.75 0h7.5C15.216 0 16 .784 16 1.75v7.5A1.75 1.75 0 0 1 14.25 11h-7.5A1.75 1.75 0 0 1 5 9.25Zm1.75-.25a.25.25 0 0 0-.25.25v7.5c0 .138.112.25.25.25h7.5a.25.25 0 0 0 .25-.25v-7.5a.25.25 0 0 0-.25-.25Z">
                    </path>
                  </svg>
                  <svg aria-hidden="true" height="16" viewBox="0 0 16 16" version="1.1" width="16" data-view-component="true"
                    class="octicon octicon-check clipboard-check-icon color-fg-success d-none m-2">
                    <path
                      d="M13.78 4.22a.75.75 0 0 1 0 1.06l-7.25 7.25a.75.75 0 0 1-1.06 0L2.22 9.28a.751.751 0 0 1 .018-1.042.751.751 0 0 1 1.042-.018L6 10.94l6.72-6.72a.75.75 0 0 1 1.06 0Z">
                    </path>
                  </svg>
                </clipboard-copy>
              </div>
            </div><p><img src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_59_0.png" alt="png" data-src="/files/UmapApplication/images/TCGA_Semisupervised_clustering_files/TCGA_Semisupervised_clustering_59_0.png" /></p></div>
        </div>

</body>
</html>
